Added image-policy proposal

erictune opened the pull request:
Add proposal for image policy.

erictune committed:
Added image-provenance proposal

erictune commented:
@wteiken

erictune commented:
@philips it might make sense to have  https://github.com/coreos/clair be a backend for this image policy webhook, or something like that?  Can you at-mention the right people from Clair?

Also @ericchiang because it is a type of authorization, but intentionally not in RBAC.

erictune commented:
@roberthbailey

ericchiang commented on docs/proposals/image-provenance.md:
```diff
@@ -0,0 +1,310 @@
+- For non-replicated things (size 1 ReplicaSet, PetSet), a single node failure may disable it.
+- a node rolling update will eventually check for liveness of replacements, and would be throttled if
+  in the case when the image was no longer allowed and so replacements could not be started.
+- rapid node restarts will cause existing pod objects to be restared by kubelet.
```
s/restared/restarted/g

erictune replied:
fixed

ericchiang commented on docs/proposals/image-provenance.md:
```diff
@@ -0,0 +1,310 @@
+* Implementing objects in core kubernetes which describe complete policies for what images are approved.
+  * A third-party implementation of an image policy checker could optionally use ThirdPartyResource to store its policy.
+* Kubernetes core code dealing with concepts of image layers, build processes, source repositories, etc.
+  * We expect there will be multiple PaaSes and/or de-facto programming enviroments, each with different takes on
```
s/enviroments/environments/

erictune replied:
fixed

erictune commented:
@jerbia

ericchiang commented:
cc @gtank @ecordell

smarterclayton commented on docs/proposals/image-provenance.md:
```diff
@@ -0,0 +1,320 @@
+ }
+ 
+// ImageReviewSpec is a description of the token authentication request.
+type ImageReviewSpec struct {
```
Do we need to know the authenticated user / service account that is attempting to perform this action?  Who can matter as much as what in this case.

smarterclayton replied:
Nm, saw the comment below.  Commented there.

smarterclayton commented on docs/proposals/image-provenance.md:
```diff
@@ -0,0 +1,320 @@
+	// Error should be empty unless Allowed is false in which case it
+        // may contain a short description of what is wrong.  Kubernetes
+        // may truncate excessively long errors when displaying to the user.
+	Error string
```
Shouldn't this be `Reason` and `Message` just like other status responses?

erictune replied:
fixed.

smarterclayton commented on docs/proposals/image-provenance.md:
```diff
@@ -0,0 +1,320 @@
+
+The ReplicaSet, or other controller, is responsible for recognizing when a 403 has happened
+(whether due to user not having permission due to bad image, or some other permission reason)
+and throttling itself and surfacing the error in a way that CLIs and UIs can show to the user.
```
I'm starting to feel like we're not doing enough to actually solve the usability problem here - the approach you describe is correct (RS needs to surface that info) but we're doing a terrible job of it.  For instance... should RS have a condition "CreationForbidden": "true" with a reason "ImageRejected"?  Should the deployment also have a condition surfaced?

Right now, we see this all the time in OpenShift (by virtue of security policy) and end users really suffer without this info being surfaced (in quota, in policy, and in crash looping pods).

erictune replied:
Yes.  https://github.com/kubernetes/kubernetes/issues/22298

kargakis replied:
> I'm starting to feel like we're not doing enough to actually solve the usability problem here - the approach you describe is correct (RS needs to surface that info) but we're doing a terrible job of it. For instance... should RS have a condition "CreationForbidden": "true" with a reason "ImageRejected"? Should the deployment also have a condition surfaced?

> Right now, we see this all the time in OpenShift (by virtue of security policy) and end users really suffer without this info being surfaced (in quota, in policy, and in crash looping pods).

Conditions are one thing that may or may not help here but I believe we already have a mechanism for reporting failures, albeit we are doing a bad job with it: events. Events are nice but fetching exactly those you are interested in is impossible currently. It feels like we should have events as a subresource to all objects or something. https://github.com/kubernetes/kubernetes/issues/11994 is related.

smarterclayton replied:
Found your answer later in the issue.  I'm forgetting which proposals I've commented on at this point.

soltysh replied:
I'd like to generalize this paragraph to be applicable to all controllers. Deplyments and ReplicaSets are not the only ones currently. But I agree with Clayton and Michalis here that we need to make it more obvious and Conditions are one of the possibilities here.

erictune replied:
Linked this discussion from #22298

smarterclayton commented on docs/proposals/image-provenance.md:
```diff
@@ -0,0 +1,320 @@
+We could have sent the username of the pod creator to the backend.  The username could be used to allow different users to run
+different categories of images.  This would require propagating the username from e.g. Deployment creation, through to
+Pod creation via, e.g. the `Impersonate-User:` header.  This feature is not ready.  When it is, we will re-evaluate
+adding user as a field of `ImagePolicyRequest`.
```
Isn't this ready now?

erictune replied:
API-server respects this header.  Controllers do not use it.

erictune replied:
Filed https://github.com/kubernetes/kubernetes/issues/27152

gtank commented on docs/proposals/image-provenance.md:
```diff
@@ -0,0 +1,320 @@
+The ImageReviewSpec is used as the key to the cache.
+
+In the case of a cache miss and timeout talking to the backend, the default is to allow Pod creation.
+Keeping services running is more important than a hypotetical threat from an un-verified image.
```
s/hypotetical/hypothetical

gtank replied:
Doesn't this mean any backend DoS lasting more than an hour will result in a fail-open cluster?

erictune replied:
Yes.  Not sure what else to do, though.

gtank replied:
Thinking out loud here: what if you fell back to using any existing "allow" cache as a whitelist if no contact with the backend can be established for a (long) period of time? This allows things that are already deployed to continue running without failing completely open.

gtank commented on docs/proposals/image-provenance.md:
```diff
@@ -0,0 +1,320 @@
+
+## Ubernetes
+
+If two clusters share an image policy backend, then they will have the same policies.
```
Is the intent to synchonrize the admission caches in the federated/ubernetes case?

erictune replied:
no.

gtank commented on docs/proposals/image-provenance.md:
```diff
@@ -0,0 +1,320 @@
+We will wait and see how much demand there is for closing this hole. If the community demands a solution,
+we may suggest one of these:
+
+1.  Use a backend that refuses to accept images that are specified with tags, and require users to resolve to IDs
```
Forcing users to resolve IDs would allow a backend to interact more effectively with content trust measures. With just the tag, there's no binding between what was actually deployed and what the backend sees. You need to somehow supply an account of the data. cc @ecordell

smarterclayton replied:
Resolving IDs before they are put into pods / deployments / rcs is the best
outcome.  Doing that resolution in the admission control is fraught (there
should be another admission controller that does the transformation, if you
really want that)

On Thu, Jun 9, 2016 at 4:53 PM, George Tankersley <notifications@github.com>
wrote:

> In docs/proposals/image-provenance.md
> <https://github.com/kubernetes/kubernetes/pull/27129#discussion_r66519606>
> :
>
> > +## Image tags and IDs
> > +
> > +Image tags are like: `myrepo/myimage:v1`.
> > +
> > +Image IDs are like: `myrepo/myimage@sha256:beb6bd6a68f114c1dc2ea4b28db81bdf91de202a9014972bec5e4d9171d90ed`.
> > +You can see image IDs with `docker images --no-trunc`.
> > +
> > +The Backend needs to be able to resolve tags to IDs (by talking to the images repo).
> > +If the Backend resolves tags to IDs, there is some risk that the tag-to-ID mapping will be
> > +modified after approval by the Backend, but before Kubelet pulls the image.  We will not address this
> > +race condition at this time.
> > +
> > +We will wait and see how much demand there is for closing this hole. If the community demands a solution,
> > +we may suggest one of these:
> > +
> > +1.  Use a backend that refuses to accept images that are specified with tags, and require users to resolve to IDs
>
> Forcing users to resolve IDs would allow a backend to interact more
> effectively with content trust measures. With just the tag, there's no
> binding between what was actually deployed and what the backend sees. You
> need to somehow supply an account of the data. cc @ecordell
> <https://github.com/ecordell>
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/27129/files/790ab4954039f8d601b925896c01962face3765f#r66519606>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABG_pykCE46gTf8z3jGttIEODyKkW9UPks5qKH1LgaJpZM4IyI6J>
> .
>

ecordell replied:
Agreed; any signature verification doesn't mean much if you don't verify that the image id matches the content

erictune replied:
I just commented on this in https://github.com/kubernetes/kubernetes/pull/27129/files/4a60be831efce93d7e210df47d79e7c18d5d13c2#r71607506

I think it makes sense to either map tag to SHA in kubectl, in CICD, or after the fact in Kubelet.  I agree the admission controller is a bad place to map tag to SHA.  I think admission is a fine place to require image names that use SHAs.

erictune commented:
xref: #22888

alex-mohr commented on docs/proposals/image-provenance.md:
```diff
@@ -0,0 +1,329 @@
+- a newly created replicaSet will be unable to create Pods.
+- updating a deployment will be safe in the sense that it will detect that the new ReplicaSet is not scaling
+  up and not scale down the old one.
+- an existing replicaSet will not be unable to create Pods that replace ones which are terminated.  If this is due to
```
nit: "will not be unable to create" -> "will be unable to create"

erictune replied:
fixed.

bgrant0607 commented on docs/proposals/image-provenance.md:
```diff
@@ -0,0 +1,329 @@
+
+## Admission controller
+
+An `ImagePolicyWebhook` admission controller will be written.  The admission controller examines all pod objects which are
```
Is the webhook going to be tried until success? How would it distinguish retryable failure from permanent failure?

erictune replied:
The admission controller will admit if the webhook times out.

Q-Lee replied:
Will the admin be able to set a policy for a namespace (e.g., fail-open/fail-closed)?

deads2k replied:
Will this be a generic webhook for any admission plugin?  I'd like to see a generic one and it seems like the work here would be about the same.

erictune replied:
It will not be a generic webhook.  A generic webhook would need a lot more discussion:

- a generic webhook needs to touch all objects, not just pods.  So it won't have a fixed schema
- a generic webhook client needs to ignore kinds it doesn't care about, or the apiserver needs to know which backends care about which kinds
- it exposes our whole API to a webhook without giving us (the project) any chance to review or understand how it is being used.
- because we don't know which fields of an object are inspected by the backend, caching is not effective.  Sending fewer fields allows caching.
- sending fewer fields makes it possible to rev the version of the webhook request slower than the version of our internal obejcts (e.g. pod v2 could still use imageReview v1.)
- probably lots more reasons.

erictune replied:
Added section about this in Alternatives section.

php-coder commented on docs/proposals/image-provenance.md:
```diff
@@ -0,0 +1,329 @@
+1.  Use a backend that refuses to accept images that are specified with tags, and require users to resolve to IDs
+    prior to creating a pod template.
+   - [kubectl could be modified to automate this process](https://github.com/kubernetes/kubernetes/issues/1697)
+   - a CI/CD system or templatingcould be used that maps IDs to tags before Deployment modification/creation.
```
s/templatingcould/templating could/

soltysh commented on docs/proposals/image-provenance.md:
```diff
@@ -0,0 +1,329 @@
+
+- request to "break glass" to override a policy, in case of emergency.
+- a ticket number from a ticket system that documents the break-glass request
+- provide a hint to the policy server as to the imageID of the image being provided, to save it a lookup
```
A good example here would be to ask the backend pick only a certain scanners, as well.

erictune replied:
Can you elaborate on this?

soltysh commented on docs/proposals/image-provenance.md:
```diff
@@ -0,0 +1,329 @@
+        // may contain a short description of what is wrong.  Kubernetes
+        // may truncate excessively long errors when displaying to the user.
+	Reason string
+}
```
I've noticed the formatting in go structs is not consistent, have a look at gh preview of this doc.

soltysh commented on docs/proposals/image-provenance.md:
```diff
@@ -0,0 +1,329 @@
+
+- never change a policy, unless some external process has ensured no active objects depend on the to-be-forbidden
+  images.
+- change a policy and assume that transition to new image happens faster than the existing pods decay.
```
With changing policy there's also the question of what to do with currently running images? Should the cluster administrator be equipped with a tool to verify which are "in danger".

erictune replied:
Good point.  The cluster admin can already list all the pods that are currently running.

What the cluster admin can't easily do is map an image name with a tag to a SHA.

Two options:

1. image provenance (or some simpler admission controller) always rejects non-SHA image names, and CI/CD is responsible for only using SHAs.  (Can use annotations or labels to store human-readable version tags, though this is sort of gross since it has to be a map from container name to tag).
2. Kubelet could post the SHA to the status of the pod after pulling it.  Seems generally useful.  Want to file that as a separate issue?   

Once there is a list of "active SHAs" in the cluster, the the cluster admin or security team or security dashboard  can then join the active SHA list with the list of SHAs that are known to have vulnerabilities.

soltysh replied:
Created https://github.com/kubernetes/kubernetes/issues/29448,

soltysh replied:
Generally, I'd prefer not to force user to use image SHA, it's quite cumbersome from a user POF, he should be able to do it (and can these days) but should not be forced. Let's face it, everybody are lazy :wink: we prefer automation wherever that is possible.

erictune replied:
Are you suggesting the apiserver map the tag to a SHA?

soltysh commented on docs/proposals/image-provenance.md:
```diff
@@ -0,0 +1,329 @@
+- Image Policy Admission Controller adds new field to Pod, e.g. `pod.spec.container[i].imageID` (or an annotation).
+  and kubelet will enforce that both the imageID and image match the image pulled.
+
+Since this adds complexity and interacts with imagePullPolicy, we avoid adding the above feature initially.
```
This definitely deserves a separate issue. I've started wondering if it would make sense to implement here OpenShift's ImageStreams but in a limited way, iow. in such a way that you could have a one-to-one mapping image to a specific image:tag which would then allow a user to specify if I care about this being a hard reference to a particular imageid at the point when that ImageStream got created or rather a following tag, iow. such that would be refreshed on every usage, which would mean it might be different each time. @smarterclayton wdyt?

erictune replied:
If you want to have a repo and build system that ensures immutability of the tags, then your image provenance backend can accept image specifiers that use tags.

soltysh commented on docs/proposals/image-provenance.md:
```diff
@@ -0,0 +1,329 @@
+Caching allows permits RC to scale up services even during short unavailability of the webhook backend.
+The ImageReviewSpec is used as the key to the cache.
+
+In the case of a cache miss and timeout talking to the backend, the default is to allow Pod creation.
```
Warning event accompanying such situation would be nice.

philips commented:
@erictune Is the plan for this to become a feature in the feature repo for v1.4? We (@ecordell) really want to help out on this.

erictune commented:
@philips ACK.  Checking with some co-workers before I answer.

erictune commented:
@philips 
Would love @ecordell to implement this.
I would like @Q-Lee to do the code reviews for @ecordell, as Q-Lee is working on a backend.

erictune commented:
I think I fixed most of the comments in my last push.  

I think @smarterclayton is pretty busy. 

 If I can get two LGTMs from the set {@soltysh, @philips, @Q-Lee, @ecordell } then I am going to take it as approval and I will merge it.

erictune commented:
@fabioy this is the proposal for image admission controller that we talked about.

philips commented on docs/proposals/image-provenance.md:
```diff
@@ -0,0 +1,334 @@
+ - only run images that are scanned to confirm they do no contain vulnerabilities
+ - only run images that use a "required" base image
+ - only run images that contain binaries which were built from peer reviewed, checked-in source
+   by a trusted compiler toolchain.
```
@erictune isn't the other obvious example images that are signed by a public key in a key list?

erictune replied:
Yes.  I would do that in conjunction with other controls.  I might want to also enforce the whitelist of signing keys at the node level, at which point the check in the apiserver is more a nice-to-have.  But you could do it that way.

erictune replied:
Adding your suggestion to the doc.

deads2k commented on docs/proposals/image-provenance.md:
```diff
@@ -0,0 +1,334 @@
+There will be a cache of decisions in the admission controller. 
+
+If the apiserver cannot reach the webhook backend, it will log a warning and admit the pod.  The rationale here is
+that keeping the system running is most important, and people who care about image provenance deeply will
```
I think I'd prefer to default the other way or at least offer an option for someone to choose which way they want to change.  If we run a bad image and it destroys something, there might not be a way to un-ring the bell.

@liggitt

liggitt replied:
+1000... getting a bad image in should not be as simple as DOSing an admission webhook endpoint

soltysh replied:
+1 the other way with an option to change it if admin wants to do so and is aware of the risk.

erictune replied:
I support flag control over deny-on-failure vs allow-on-failure. 

Doing that per-namespace, as I think @q-lee mentioned above, would not work well with a flag.  @q-lee do you have a use case for per-namespace?

erictune replied:
Interesting question: how to bootstrap image provenance controller running on the cluster itself.

erictune replied:
Adding line about the flag.

smarterclayton commented on docs/proposals/image-provenance.md:
```diff
@@ -0,0 +1,334 @@
+An `ImagePolicyWebhook` admission controller will be written.  The admission controller examines all pod objects which are
+created or updated.  It can either admit the pod, or reject it.  If it is rejected, the request sees a `403 FORBIDDEN`
+
+The admission controller code will go in `plugin/pkg/admission/imagepolicy`.
```
I would like for the admission controller to take an interface for "check decision" rather than embedding all the client logic in it.  That would allow alternate admission controller implementations to be more easily implemented.  As an example, we're trying to build composable chunks of admission logic for policy like this that can be reused in other contexts.  Things like authorizer and authentication have succeeded pretty well at this (@liggitt has done some crazy authenticator wrappers that work cleanly).  I would like to generally have our policy decision steps behind clean interfaces (in this case, an interface that answers the question about an image and it being accepted and mirrors the ImageReview object) that can be composed later on.

deads2k replied:
> I would like for the admission controller to take an interface for "check decision" rather than embedding all the client logic in it.

You're saying you don't like the webhook mechanism suggested, that you want a generic webhook like I asked above above, or that you want the reference webhook impl to accept what amounts to and admission.Interface?

smarterclayton replied:
On Jul 25, 2016, at 3:21 PM, David Eads <notifications@github.com> wrote:

In docs/proposals/image-provenance.md
<https://github.com/kubernetes/kubernetes/pull/27129#discussion_r72127227>:

> +    reduces latency and allows short outages of the backend to be tolerated.
> +
> +
> +Detailed discussion in [Ensuring only images are from approved sources are run](
> +https://github.com/kubernetes/kubernetes/issues/22888).
> +
> +# Implementation
> +
> +A new admission controller will be added.  That will be the only change.
> +
> +## Admission controller
> +
> +An `ImagePolicyWebhook` admission controller will be written.  The admission controller examines all pod objects which are
> +created or updated.  It can either admit the pod, or reject it.  If it is rejected, the request sees a `403 FORBIDDEN`
> +
> +The admission controller code will go in `plugin/pkg/admission/imagepolicy`.

I would like for the admission controller to take an interface for "check
decision" rather than embedding all the client logic in it.

You're saying you don't like the webhook mechanism suggested, that you want
a generic webhook like I asked above above, or that you want the reference
webhook impl to accept what amounts to and admission.Interface?


The latter.  Authorizer and Admission interfaces are very successful.
Would like to try to define the equivalents for other policy equally well.

smarterclayton commented on docs/proposals/image-provenance.md:
```diff
@@ -0,0 +1,334 @@
+// ImageReviewSpec is a description of the pod creation request.
+type ImageReviewSpec struct {
+	// Containers is a list of a subset of the information in each container of the Pod being created.
+	Containers []ImageReviewContainerSpec
```
Unfortunately you also need to accept init containers.  I would recommend creating a nested struct that is similar to pod template that has the subset of info in it, and make that hierarchal rather than flattened, i.e.:

```
type ImageReviewPodTemplate struct {
  Metadata ImageReviewObjectMeta
  Spec ImageReviewPodSpec
}
```

If we can preserve the same hierarchy as a `PodTemplate`, that makes automated tools easier in the future (we might take the pod object as unstructured and whitelist the things that are included in a generic fashion).

smarterclayton replied:
It also establishes a pattern for future examples of this where subsets of data are returned.

erictune replied:
Is it settled where in the PodSpec that init containers will live?

erictune replied:
Also, if we do what you said, it couples the versioning of Pods to the versioning of the ImageReview api.  Is that desirable?

smarterclayton replied:
If is extremely likely init containers will be a peer to containers,
barring any sudden revelations.

smarterclayton replied:
Possibly not, although creating a new structure to learn *is* causing
API drift.  I would expect this API to evolve to be consistent with
pods in the long term (v1 pods to vX ImagePolicyReview)

Q-Lee replied:
I'm of the opinion that instead of mimicking pod layout, we should eliminate the container list altogether. The goal here is to establish a chain of trust for containers, and not to enforce pod-level policies. If you create a pod with 5 containers, then you make 5 requests to the backend.

deads2k replied:
> I'm of the opinion that instead of mimicking pod layout, we should eliminate the container list altogether. The goal here is to establish a chain of trust for containers, and not to enforce pod-level policies. If you create a pod with 5 containers, then you make 5 requests to the backend.

I think that it's likely we'll end up doing both.  Given the current power of PSP and it's ability to describe who can request which powers for a pod/container, it seems likely that the decision about whether a particular image is allowed may be affected by the same or a similar policy.

I'm not suggesting that we add that level of complication now, but designing the structure for future expansion seems like a reasonable thing to do and this would be a way to do it.

Also, remote callouts are expensive and if we already have all the data ready, validating it all at once seems pretty reasonable.

Q-Lee replied:
Exactly, we have PSP for pod level control. The purpose here is to establish a chain of trust from source code to deployment.

A small scaling factor from pods to containers is nothing to bat an eye at.

soltysh commented on docs/proposals/image-provenance.md:
```diff
@@ -0,0 +1,334 @@
+
+The exact nature of "approval" is beyond the scope of Kubernetes, but may include reasons like:
+
+ - only run images that are scanned to confirm they do no contain vulnerabilities
```
s/do no/do not

soltysh commented on docs/proposals/image-provenance.md:
```diff
@@ -0,0 +1,334 @@
+* Block creation of pods that would cause "unapproved" images to run.
+* Make it easy for users or partners to build "image provenance checkers" which check whether images are "approved".
+  *  We expect there will be multiple implementations.
+* Allow users to request an "override" of the policy in a convenient way (subject to the override being allowed).
```
Is the override be available to all users? Or this will be tied to specific authz?

erictune replied:
In one possible implementation, the override is available to all users, but a user who requests the override would be expected to answer to an auditor, sometime after the fact, if she requests the override.

erictune commented:
I've addressed most comments and updated the docs.  

The open issues, as I see it, are:
- how to handle init-containers
- if the schema should be a subset of PodSpec, so that additional fields fall in the same places.
- something about interfaces that I don't quite follow.

smarterclayton commented:
Interfaces was about making the guts of the admission controller call
a policy interface that looks like Authorizer and Admission
interfaces, because it imposes discipline and also may lead to
experimentation.

philips commented:
cc the Clair team @Quentin-M @jzelinskie @josephschorr

jzelinskie commented on docs/proposals/image-provenance.md:
```diff
@@ -0,0 +1,358 @@
+## Admission controller
+
+An `ImagePolicyWebhook` admission controller will be written.  The admission controller examines all pod objects which are
+created or updated.  It can either admit the pod, or reject it.  If it is rejected, the request sees a `403 FORBIDDEN`
```
This description is a little vague in the plurality sense. I think this should work over a set of webhooks rather than one. For example, this Pull Request on GitHub has multiple webhooks that must validate it before it is merged.

soltysh replied:
I kinda understood this as being able to setup multiple, but having it explicit in the proposal is a good idea.

deads2k replied:
> This description is a little vague in the plurality sense. I think this should work over a set of webhooks rather than one. For example, this Pull Request on GitHub has multiple webhooks that must validate it before it is merged.

Seems like we could support a single callout and if someone wanted a union, they could write the union in their particular handler.  That keeps our core code out of the business of deciding between the ands, ors, and trumps, which inevitably follow the "give me more than one".

I don't see an issue with making a reference impl for the webhook that can provide a simple union, but I don't think we want to bake multiples into our admission plugin.

soltysh replied:
I feel convinced.

jzelinskie replied:
@deads2k It seems there's precedent for doing it the way you have described in other places, so I agree.

ericchiang commented on docs/proposals/image-provenance.md:
````diff
@@ -0,0 +1,358 @@
+The WebHook request and response are JSON, and correspond to the following `go` structures:
+
+```go
+// Filename: pkg/apis/authentication.k8s.io/register.go
````
Are these types intended to live in the `authentication.k8s.io` API group? Is the package name correct here?

erictune replied:
No, fixed.

Q-Lee commented on docs/proposals/image-provenance.md:
```diff
@@ -0,0 +1,358 @@
+
+// ImageReviewContainerSpec is a description of a container within the pod creation request.
+type ImageReviewContainerSpec struct {
+	Image string
```
Shouldn't this be Image, ImageHash string?

erictune replied:
Images can be specified to docker, and in `pod.spec.container[].image` as either `image:tag` or `image@SHA:012345679abcdef`.  So, this field also accepts either format.

It is up to the backend to decide if it accepts `image:tag` or only accepts `image@SHA:012345679abcdef` format.   There are reasons you might chose to do either way, so this API doesn't have an opinion.

Q-Lee commented:
This is very close. Let's put this and I'll make the remaining changes in a new, narrower PR.
